### ISSUE TITLE
Unable to fetch cross-instance posts

### ISSUE THREAD

**Issue by Kradyz** [LemmyNet/lemmy#1994]
### Issue Summary

If a user from instance A makes a post to a community in instance B, this post cannot be fetched by instance C. This appears to be an issue only with instances that have upgraded to 0.14.5

### Steps to Reproduce

Find a post in lemmy.ml posted by a user in a different instance that has upgraded to version 0.14.5. Attempt to fetch the post from a third instance, or use a curl command. Both the lemmy.ml url and the original instance's URL will fail to fetch. The original instance URL will succeed to fetch if the instance is still using version 0.14.3.

Examples:

Post from feddit.de (version 0.14.5) to lemmy.ml/c/antiwork:

curl -H 'Accept: application/activity+json' https://lemmy.ml/post/113699
curl -H 'Accept: application/activity+json' https://feddit.de/post/25493

Both fail.

Post from midwest.social  (version 0.14.5)  to lemmy.ml/c/artificial_intel:

curl -H 'Accept: application/activity+json' https://lemmy.ml/post/113652
curl -H 'Accept: application/activity+json' https://midwest.social/post/14276

Both fail.

An example from lemmy.ca (version 0.14.5) to lemmy/ml/c/lemmy_support:

curl -H 'Accept: application/activity+json' https://lemmy.ml/post/107664
curl -H 'Accept: application/activity+json' https://lemmy.ca/post/26505

Both fail. However, case 2 fails in a different way, returning "NotFound" and nothing else. This one I can actually access from my instance  (https://mander.xyz/post/34). I think that I might have fetched these before the 0.14.5 upgrade. 

Sopuli is still in version 0.14.3, posts can be fetched directly from sopuli but the same post can't be fetched from lemmy.ml:

Lemmy URLs:
curl -H 'Accept: application/activity+json' https://lemmy.ml/post/113150
curl -H 'Accept: application/activity+json' https://lemmy.ml/post/84053

Sopuli URLs:
curl -H 'Accept: application/activity+json'  https://sopuli.xyz/post/23384
curl -H 'Accept: application/activity+json' https://sopuli.xyz/post/18318

### Technical details

The curl error is the following:

```

NotFound   0: tracing_actix_web::root_span_builder::HTTP request
           with http.method=GET http.route=/post/{post_id} http.flavor=1.0 http.scheme=http http.host=lemmy.ml http.client_ip=xx.xx.xx.xx http.user_agent=curl/7.68.0 http.target=/post/84053 otel.kind="server" request_id=0ddba244-2129-4288-be72-67eddfe6cf63
             at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-actix-web-0.5.0-beta.2/src/root_span_builder.rs:40root@ubuntu2004:/var/www/mander#

```

When failing to fetch one of these posts from my instance, logs will show the following error:
```

lemmy_1     |

lemmy_1     |   2021-12-13T11:30:45.750988Z  INFO lemmy_apub_lib::object_id: Fetching remote object https://lemmy.ml/post/84053

lemmy_1     |     at crates/apub_lib/src/object_id.rs:115

lemmy_1     |

lemmy_1     |   2021-12-13T11:30:45.773262Z  WARN lemmy_utils: error decoding response body: expected value at line 1 column 1

lemmy_1     |     at crates/utils/src/lib.rs:60

lemmy_1     |

lemmy_1     |   2021-12-13T11:30:45.774847Z ERROR lemmy_websocket::handlers: Error during message handling {"error":"couldnt_find_object"}

lemmy_1     |     at crates/websocket/src/handlers.rs:78

lemmy_1     |

```

**Comment by dessalines** [comment]
I don't think this is a bug? You should be using the original post URL, where the post actually lives. Otherwise you start building long request chains.

This is also how inbox pushing works, it doesn't chain them.

**Comment by Kradyz** [comment]
To focus on this post specifically:

curl -H 'Accept: application/activity+json' https://lemmy.ml/post/113652
curl -H 'Accept: application/activity+json' https://midwest.social/post/14276

Curl 1 should fail, but curl 2 should succeed, because the post lives in midwest.social, as I understand it. But both curls fail, and I am not able to fetch it from my instance. It is as if since a poster from midwest.social posted it to lemmy.ml, then it does not live in either.

**Comment by Nutomic** [comment]
You are right, there seems to be something wrong with that specific post on midwest.social. Others are working fine though. Would need the instance admin to look into.

**Comment by Kradyz** [comment]
Hmm, ok! But 
I did do multiple tests and the same happened for every post  tested when the post was made from an instance with version 0.14.5:

This one is actually posted from a lemmy user to feddi, so the lemmy.ml curl should work:

curl -H 'Accept: application/activity+json' https://lemmy.ml/post/116413
curl -H 'Accept: application/activity+json' https://feddit.de/post/27147

Another example:
curl -H 'Accept: application/activity+json' https://fapsi.be/post/15703
curl -H 'Accept: application/activity+json' https://lemmy.ml/post/114422


Then, I reverted my image of Lemmy to 0.14.3, and all of the posts that had been posted from Mander previously now do work. 

So, this one does curl properly:
curl -H 'Accept: application/activity+json' https://mander.xyz/post/2445

I then switched the image 0.14.5 again and made this post, which again fails:
curl -H 'Accept: application/activity+json' https://mander.xyz/post/2497
curl -H 'Accept: application/activity+json' https://enterprise.lemmy.ml/post/494
 
I don't understand what migrations do exactly, and whether they are executed only at a point in time or whether they are normal code, so this might not be a good guess, but: looking at changes between 0.14.5 and 0.14.3, the "Delete dupes" migration stood out to me: migrations/2021-11-22-135324_add_activity_ap_id_index/up.sql
Could this be considering the cross-instance post a dupe and removing its id from the activities?

**Comment by Nutomic** [comment]
So your problem is that posts from your instance dont get delivered to another instance? Can you check if there is anything relevant in the logs?

**Comment by Kradyz** [comment]
I can replicate this issue using only the lemmy.ml instances. That might help clarify the issue.

You will not be able to fetch this post from lemmy.ml:

https://enterprise.lemmy.ml/post/541

This post was made from a user in the "Enterprise" instance into a community in the "Voyager" instance. 

If I try to fetch it from my instance, the docker logs show the "couldnt_find_object" error that I posted above.

**Comment by Kradyz** [comment]
The issue is that I am not able to fetch posts that were posted from poster of instance A to a community in instance B. This is not specific to my instance, I have tried it with every instance I could find.

I did manage to find a single counter-example. I made a post in a community of voyager.lemmy.ml using a user from enterprise.lemmy.ml, and it was possible to fetch the post from ds9.lemmy.ml.

This is the post: https://enterprise.lemmy.ml/post/541

I have tried to fetch the post from every other instance I could find, and it is not possible to fetch it.


This is the Docker log when I try to fetch it:

`lemmy_1     |   2021-12-16T21:20:07.225617Z  INFO lemmy_websocket::routes: Message received: "{\"op\":\"ListCommunities\",\"data\":{\"type_\":\"All\",\"sort\":\"TopAll\",\"limit\":20,\"auth\":\"--------\"}}" from id: 3059251804329081406
lemmy_1     |     at crates/websocket/src/routes.rs:118
lemmy_1     |     in tracing_actix_web::root_span_builder::HTTP request with , http.method: GET, http.route: /api/v3/ws, http.flavor: 1.1, http.scheme: http, http.host: mander.xyz, http.client_ip: xx.xx.xx.xx, http.user_agent: Mozilla/5.0 (X11; Linux x86_64; rv:95.0) Gecko/20100101 Firefox/95.0, http.target: /api/v3/ws, otel.kind: "server", request_id: beb073ed-3842-4759-838a-1af7727cc36c, http.status_code: 101, otel.status_code: "OK"
lemmy_1     |
lemmy_1     |   2021-12-16T21:20:07.230071Z  INFO lemmy_websocket::routes: Message received: "{\"op\":\"Search\",\"data\":{\"q\":\"https://enterprise.lemmy.ml/post/541\",\"type_\":\"All\",\"sort\":\"TopAll\",\"listing_type\":\"All\",\"page\":1,\"limit\":20,\"auth\":\"-------\"}}" from id: 3059251804329081406
lemmy_1     |     at crates/websocket/src/routes.rs:118
lemmy_1     |     in tracing_actix_web::root_span_builder::HTTP request with , http.method: GET, http.route: /api/v3/ws, http.flavor: 1.1, http.scheme: http, http.host: mander.xyz, http.client_ip: xx.xxx.xxx.xxx, http.user_agent: Mozilla/5.0 (X11; Linux x86_64; rv:95.0) Gecko/20100101 Firefox/95.0, http.target: /api/v3/ws, otel.kind: "server", request_id: beb073ed-3842-4759-838a-1af7727cc36c, http.status_code: 101, otel.status_code: "OK"
lemmy_1     |
lemmy_1     |   2021-12-16T21:20:07.232280Z  INFO lemmy_websocket::routes: Message received: "{\"op\":\"ResolveObject\",\"data\":{\"q\":\"https://enterprise.lemmy.ml/post/541\",\"auth\":\"-------\"}}" from id: 3059251804329081406
lemmy_1     |     at crates/websocket/src/routes.rs:118
lemmy_1     |     in tracing_actix_web::root_span_builder::HTTP request with , http.method: GET, http.route: /api/v3/ws, http.flavor: 1.1, http.scheme: http, http.host: mander.xyz, http.client_ip: xx.xxx.xxx.xxx, http.user_agent: Mozilla/5.0 (X11; Linux x86_64; rv:95.0) Gecko/20100101 Firefox/95.0, http.target: /api/v3/ws, otel.kind: "server", request_id: beb073ed-3842-4759-838a-1af7727cc36c, http.status_code: 101, otel.status_code: "OK"
lemmy_1     |
lemmy_1     |   2021-12-16T21:20:07.242345Z  INFO lemmy_apub_lib::object_id: Fetching remote object https://enterprise.lemmy.ml/post/541
lemmy_1     |     at crates/apub_lib/src/object_id.rs:115
lemmy_1     |
lemmy_1     |   2021-12-16T21:20:07.281519Z  WARN lemmy_utils: error decoding response body: expected value at line 1 column 1
lemmy_1     |     at crates/utils/src/lib.rs:60
lemmy_1     |
lemmy_1     |   2021-12-16T21:20:07.282478Z ERROR lemmy_websocket::handlers: Error during message handling {"error":"couldnt_find_object"}
lemmy_1     |     at crates/websocket/src/handlers.rs:78
`

The fact that the enterprise -> voyager test post it can be pulled from ds9.lemmy.ml suggests a few possibilities. One of them is that whatever is causing the issue is already fixed in v0.15, in which case the issue will be fixed with the next update.

**Comment by Nutomic** [comment]
That one is broken on enterprise.lemmy.ml:
```
curl -H 'Accept: application/activity+json' https://enterprise.lemmy.ml/post/541
NotFound^J      
```

Logs essentially just say `NotFound`, no idea why. It looks normal in the database.

**Comment by Kradyz** [comment]
Whatever caused the issue, it appears to be fixed when updating lemmy to v0.15, so I think that the issue can be safely closed.
